packages/lib/src/validateChildren.tsx

```diff
@@ -0,0 +1,41 @@
+import isUndefined from 'lodash/isUndefined';
+import React, { ReactElement } from 'react';
+import { consoleOnce, isComponentType } from '.';
+
+/**
+ * Filters children down to a restricted set of component types.
+ *
+ * Example:
+ * ```ts
+ * // `options` will only include `ComboboxOption` components,
+ * // and log an error to the console for any invalid children
+ * const options = validateChildren(children, ['ComboboxOption'])
+ * ```
+ *
+ * @returns Array<ReactElement>
+ */
+export const validateChildren = (
+  /** Any React children */
+  children: React.ReactNode,
+  /** Array of `displayNames` for valid components */
+  validTypes: Array<string>,
+): Array<ReactElement> | undefined => {
+  const validChildren = React.Children.map(children, child => {
+    if (validTypes.some(type => isComponentType(child, type))) {
+      return child as React.ReactElement;
+    }
+  })?.filter(child => !isUndefined(child));
```

Comment:
nit: looks like this is running two loops (`map` and `filter`) when it can be run in one?

Reply:
I guess a `reduce` loop would work here, yeah